### PR TITLE
feature/dev-292/newsfeed

### PR DESCRIPTION
**Pull request by RamsesDijkstra**
First PR to unlock frontend..
Contains activity memory and db layers and a few endpoints

### CODE REVIEW

File: universe/activities/activities.go [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+	"github.com/momentum-xyz/ubercontroller/utils/umid"
+)
+
+var _ universe.Activities = (*Activities)(nil)
+
+type Activities struct {
+	ctx        types.NodeContext
+	log        *zap.SugaredLogger
+	db         database.DB
+	activities *generic.SyncMap[umid.UMID, universe.Activity]
```
Comment (jor-rit) [inline universe/activities/activities.go:24]: hrm, same problem I had with the assets implementation. Probably an exponentially growing list, in memory where only the 'last' couple of entries are frequently used :/
So I would have kept it in DB, but I guess i'm too late for that again.

File: universe/activities/activities.go [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+}
+
+func (a *Activities) GetActivity(activityID umid.UMID) (universe.Activity, bool) {
+	asset, ok := a.activities.Load(activityID)
+	return asset, ok
+}
+
+func (a *Activities) GetActivities() map[umid.UMID]universe.Activity {
+	a.activities.Mu.RLock()
+	defer a.activities.Mu.RUnlock()
```
Comment (jor-rit) [inline universe/activities/activities.go]: Again, not a blocker, but SyncMap also implements it own locking, so it's kindof duplicate here.
(e.g. `a.activities.Map`)

File: universe/activities/api.go [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package activities
+
+import (
+	"github.com/gin-gonic/gin"
+)
+
+func (a *Activities) RegisterAPI(r *gin.Engine) {
+	a.log.Info("Registering api for activities...")
```
Comment (jor-rit) [inline universe/activities/api.go]: not used, so can just be deleted

File: universe/node/api.go [hunk 4]
```diff
@@ -152,6 +148,12 @@ func (n *Node) RegisterAPI(r *gin.Engine) {
 				object.GET("/attributes/sub", n.apiGetObjectAttributeSubValue)
 
 				object.GET("/all-users/attributes", n.apiGetObjectAllUsersAttributeValuesList)
+
+				timeline := object.Group("/timeline")
+				{
+					timeline.GET("", n.apiTimelineForObject)
+					timeline.POST("", n.apiTimelineAddForObject)
```
Comment (jor-rit) [inline universe/node/api.go:155]: suspect the POST will need some additional permission checking wrapper around it

**Review by jor-rit — APPROVED** [review]
OK, to unblock frontend devs :)
Do have some question marks, but that can be fixed later

**Review by dmitry-yudakov — APPROVED** [review]
I think it's good initial step that will unblock the FE dev, but it should be optimised in future:

- GET activity currently calling `GetPaginatedActivitiesByObjectID` will be the most commonly used - it shouldn't do the paging/sorting/etc every time.
- page/pageSize is not too good approach for our case of often changed data and could lead to duplication - it's better be olderThanId(or Timestamp) and max size. We could also pass the last known id/timestamp and check for posts newer than that, to be able to show indication.
- like @jor-rit mentioned, it's not too good approach to load this constantly growing and potentially very large data into memory - we could either rely entirely on DB or has some cache in the app memory to make ADD/GET operations fast and efficient.
- We should rethink the idea of using the activity log for user posts. I'd argue that activity log should be only the log of activities, not the storage of info like posts, events, etc. It shouldn't have UPDATE operation - but rather update something outside the log and put the event about update into the log.
- This whole module is IMO different from other modules (or at least asset-3d that I'm mostly familiar with) and probably could be structured differently.

**Pull request merged**